sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -403,18 +396,75 @@ class GpuSubPartitionPairIterator(
   }
 
   private[this] def tryPullNextPair(): Option[PartitionPair] = {
-    if(hasNextBatch()) {
-      val (partIds, spillBuildBatch) = buildSubIterator.next()
-      closeOnExcept(spillBuildBatch) { _ =>
-        closeOnExcept(ArrayBuffer.empty[SpillableColumnarBatch]) { streamBuf =>
-          partIds.foreach { id =>
-            streamBuf ++= streamSubPartitioner.releaseBatchesByPartition(id)
+    var pair: Option[PartitionPair] = None
+    var continue = true
+    while(continue) {
+      if (hasNextBatch()) {
+        val (partIds, buildBatches) = buildSubIterator.next()
+        val (buildPartsSize, streamBatches) = closeOnExcept(buildBatches) { _ =>
+          val batchesSize = buildBatches.map(_.sizeInBytes).sum
+          closeOnExcept(ArrayBuffer.empty[SpillableColumnarBatch]) { streamBuf =>
+            partIds.foreach { id =>
+              streamBuf ++= streamSubPartitioner.releaseBatchesByPartition(id)
+            }
+            (batchesSize, streamBuf)
           }
-          Some(new PartitionPair(spillBuildBatch, streamBuf))
         }
+        closeOnExcept(streamBatches) { _ =>
+          if (!repartitioned && buildPartsSize > targetBatchSize) {
+            // Got a partition the size is larger than the target size. Cache it and
+            // its corresponding stream batches.
+            closeOnExcept(buildBatches)(bigBuildBatches ++= _)
+            bigStreamBatches ++= streamBatches
+          } else {
+            // Got a normal pair, return it
+            continue = false
+            val buildBatch = GpuSubPartitionHashJoin.concatSpillBatchesAndClose(buildBatches)
+            pair = Some(new PartitionPair(buildBatch, streamBatches))
+          }
+        }
+      } else if (bigBuildBatches.nonEmpty) {
```

Comment:
the way this seems to be working is that big batches are handled later, after the original sub partitioner iterators are flushed. Do I understand that correctly? If so I wonder if it makes sense to stop pulling from the sub partitioner, and give priority to the big batches temporarily, because that should alleviate the need to spill all the batches in `bigBuildBatches`.

Reply:
This sounds good to me. I filed https://github.com/NVIDIA/spark-rapids/issues/8057